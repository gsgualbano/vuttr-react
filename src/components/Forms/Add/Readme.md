#### FormulÃ¡rio para adicionar novas ferramentas

Exemplo:

```js
const setShowModal = () => {};

<Add setShowModal={setShowModal} />;
```
