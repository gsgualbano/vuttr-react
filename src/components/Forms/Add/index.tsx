import React, { useState } from 'react';
import { Form } from '@rocketseat/unform';
import * as Yup from 'yup';

import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import * as ToolsActions from '~/store/ducks/tools/actions';
import { FormData } from '~/store/ducks/tools/types';

import {
  Container,
  Header,
  Content,
  TextInput,
  Bottom,
  Label,
  TextArea,
  FormGroup,
  IconContainer,
  PlusIcon,
} from './styles';
import TagsInput from '~/components/TagsInput';
import PrimaryButton from '~/components/Common/PrimaryButton';

const schema = Yup.object().shape({
  title: Yup.string().required('Title required'),
  description: Yup.string().required('Description required'),
  link: Yup.string().required('Link required'),
});

/**
 * Props que vem do redux
 */
interface DispatchProps {
  /** ação para adicionar uma nova ferramena a api */
  addTool(data: FormData): void;
}

/**
 * Props do AddForm
 */
interface OwnProps {
  /** dispatcher para mostrar o modal */
  setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
}

type Props = DispatchProps & OwnProps;

/**
 * Formulário de adição de novas ferramentas
 */
const AddForm = ({ addTool, setShowModal }: Props): JSX.Element => {
  const [tags, setTags] = useState<string[]>([]);

  const handleSubmit = (data: any) => {
    const tool: FormData = {
      title: data.title,
      link: data.link,
      description: data.description,
      tags,
    };

    addTool(tool);
    setShowModal(false);
  };

  return (
    <Container>
      <Header>
        <IconContainer>
          <PlusIcon />
        </IconContainer>
        Add new tool
      </Header>
      <Form schema={schema} onSubmit={handleSubmit}>
        <Content>
          <FormGroup>
            <Label htmlFor="title">Tool Name</Label>
            <TextInput data-testid="title" id="title" name="title" placeholder="Name" />
          </FormGroup>
          <FormGroup>
            <Label htmlFor="link">Tool Link</Label>
            <TextInput data-testid="link" id="link" name="link" placeholder="Link" />
          </FormGroup>
          <FormGroup>
            <Label htmlFor="description">Tool description</Label>
            <TextArea
              data-testid="description"
              id="description"
              name="description"
              placeholder="Description"
              rows={6}
            />
          </FormGroup>
          <FormGroup>
            <Label htmlFor="tags">Tags</Label>
            <TagsInput id="tags" name="tags" tags={tags} setTags={setTags} />
          </FormGroup>
        </Content>
        <Bottom>
          <PrimaryButton type="submit">Add tool</PrimaryButton>
        </Bottom>
      </Form>
    </Container>
  );
};

const mapDispatchToProps = (dispatch: Dispatch) => bindActionCreators(ToolsActions, dispatch);

export default connect(
  null,
  mapDispatchToProps,
)(AddForm);
