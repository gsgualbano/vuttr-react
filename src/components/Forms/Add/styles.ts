import styled from 'styled-components';
import { Input, Textarea } from '@rocketseat/unform';
import { ReactComponent as Icon } from '~/assets/icons/Icon-Plus-Circle-2px.svg';

export const Container = styled.div``;

export const Header = styled.div`
  display: flex;
  flex-direction: row;
  margin: 10px 0 20px;
`;

export const Content = styled.div``;

export const FormGroup = styled.div`
  margin: 10px 0;

  span {
    font-size: 14px;
    color: #f95e5a;
  }
`;

export const Label = styled.label`
  font-size: 14px;
`;

export const TextInput = styled(Input)`
  display: block;
  border-radius: 5px;
  border: 1px solid;
  border-color: ${({ theme }) => theme.colors.InputBorderColor};
  background: ${({ theme }) => theme.colors.InputBackground};
  width: 95%;
  padding: 10px;
  font-family: 'Source Sans Pro, Regular', sans-serif;
`;

export const TextArea = styled(Textarea)`
  display: block;
  border-radius: 5px;
  border: 1px solid;
  border-color: ${({ theme }) => theme.colors.InputBorderColor};
  background: ${({ theme }) => theme.colors.InputBackground};
  width: 95%;
  padding: 10px;
  font-family: 'Source Sans Pro, Regular', sans-serif;
`;

export const Bottom = styled.div`
  text-align: right;
`;

export const IconContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const PlusIcon = styled(Icon)`
  width: 18px;
  height: 18px;
  margin-right: 10px;

  g > circle.a {
    stroke: #12db89;
  }

  g > path.b {
    stroke: #12db89;
    stroke-width: 5px;
  }
`;
