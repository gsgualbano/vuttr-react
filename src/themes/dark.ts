export default {
  colors: {
    background: '#1B1A1A',
    textColor: '#fff',
    cardBackground: '#29292B',
    cardBorderColor: '#29292B',
    modalBackground: '#29292B',
    modalBorderColor: '#29292B',
    linkColor: '#ae8fd6',
    linkVisitedColor: '##8050bf',
    buttonOutlineBackground: '#29292B',
    InputBackground: '#9e9fa2',
    InputBorderColor: '#9e9fa2',
  },
};
