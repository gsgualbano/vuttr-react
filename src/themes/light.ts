export default {
  colors: {
    background: '#fcfcfd',
    textColor: '#170c3a',
    cardBackground: '#fcfcfd',
    cardBorderColor: '#ebeaed',
    modalBackground: '#fefefe',
    modalBorderColor: '#888',
    linkColor: 'rgb(0, 0, 238)',
    linkVisitedColor: 'rgb(85, 26, 139);',
    buttonOutlineBackground: '#ffffff',
    InputBackground: '#f5f4f6',
    InputBorderColor: '#ebeaed',
  },
};
