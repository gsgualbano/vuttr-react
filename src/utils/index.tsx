/* eslint-disable max-len */
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { reducer } from '../store/ducks/tools';

interface Theme {
  colors: {
    background: string;
    textColor: string;
    cardBackground: string;
    cardBorderColor: string;
    modalBackground: string;
    modalBorderColor: string;
    linkColor: string;
    linkVisitedColor: string;
    buttonOutlineBackground: string;
    InputBackground: string;
    InputBorderColor: string;
  };
}

/** redux mock to render components that depends on the redux */
export const renderWithRedux = (
  component: JSX.Element,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(<Provider store={store}>{component}</Provider>),
  store,
});

/** render components thats depends on the theme provider */
export const renderWithTheme = (component: JSX.Element, theme: Theme) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

/** render components thats depends on the redux and theme provider */
export const renderWithReduxAndTheme = (
  component: JSX.Element,
  theme: Theme,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>{component}</Provider>
    </ThemeProvider>,
  ),
  store,
});
