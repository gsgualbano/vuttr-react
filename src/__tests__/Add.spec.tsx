import React from 'react';
import { renderWithReduxAndTheme } from '~/utils';
import { cleanup } from '@testing-library/react';

import Add from '~/components/Forms/Add';

import 'jest-styled-components';

import light from '~/themes/light';
import dark from '~/themes/dark';

afterEach(cleanup);

describe('<Add />', () => {
  it('render correctly', () => {
    const { asFragment } = renderWithReduxAndTheme(<Add setShowModal={() => {}} />, light);
    expect(asFragment()).toMatchSnapshot();
  });

  it('have correct light theme', () => {
    const { getByTestId } = renderWithReduxAndTheme(<Add setShowModal={() => {}} />, light);
    expect(getByTestId('title')).toHaveStyleRule('background', '#f5f4f6');
    expect(getByTestId('link')).toHaveStyleRule('background', '#f5f4f6');
    expect(getByTestId('description')).toHaveStyleRule('background', '#f5f4f6');
  });

  it('have correct dark theme', () => {
    const { getByTestId } = renderWithReduxAndTheme(<Add setShowModal={() => {}} />, dark);
    expect(getByTestId('title')).toHaveStyleRule('background', '#9e9fa2');
    expect(getByTestId('link')).toHaveStyleRule('background', '#9e9fa2');
    expect(getByTestId('description')).toHaveStyleRule('background', '#9e9fa2');
  });
});
