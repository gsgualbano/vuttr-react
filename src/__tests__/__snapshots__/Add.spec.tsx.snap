// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Add /> render correctly 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin: 10px 0 20px;
}

.c3 {
  margin: 10px 0;
}

.c3 span {
  font-size: 14px;
  color: #f95e5a;
}

.c4 {
  font-size: 14px;
}

.c5 {
  display: block;
  border-radius: 5px;
  border: 1px solid;
  border-color: #ebeaed;
  background: #f5f4f6;
  width: 95%;
  padding: 10px;
  font-family: 'Source Sans Pro, Regular',sans-serif;
}

.c6 {
  display: block;
  border-radius: 5px;
  border: 1px solid;
  border-color: #ebeaed;
  background: #f5f4f6;
  width: 95%;
  padding: 10px;
  font-family: 'Source Sans Pro, Regular',sans-serif;
}

.c9 {
  text-align: right;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  width: 18px;
  height: 18px;
  margin-right: 10px;
}

.c2 g > circle.a {
  stroke: #12db89;
}

.c2 g > path.b {
  stroke: #12db89;
  stroke-width: 5px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  min-width: 100%;
  border: 1px solid;
  border-color: #ebeaed;
  background: #f5f4f6;
  border-radius: 5px;
}

.c8 {
  background: inherit;
  border: none;
  padding: 10px;
  outline: none;
  width: 100%;
}

.c10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: 16px;
  color: #ffffff;
  background: #365df0;
  padding: 5px 10px;
  border-radius: 5px;
  border: none;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c10:hover {
  background: #2f55cc;
}

.c10:click {
  background: #244aa8;
  background-color: #244aa8;
}

@media (max-width:768px) {
  .c10 {
    width: 100%;
  }
}

<div
    class=""
  >
    <div
      class="c0"
    >
      <div
        class="c1"
      >
        <svg
          class="c2"
        >
          Icon-Plus-Circle-2px.svg
        </svg>
      </div>
      Add new tool
    </div>
    <form
      data-testid="form"
    >
      <div
        class=""
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for="title"
          >
            Tool Name
          </label>
          <input
            aria-label="title"
            class="c5"
            data-testid="title"
            id="title"
            name="title"
            placeholder="Name"
            value=""
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="link"
          >
            Tool Link
          </label>
          <input
            aria-label="link"
            class="c5"
            data-testid="link"
            id="link"
            name="link"
            placeholder="Link"
            value=""
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="description"
          >
            Tool description
          </label>
          <textarea
            aria-label="description"
            class="c6"
            data-testid="description"
            id="description"
            name="description"
            placeholder="Description"
            rows="6"
          />
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for="tags"
          >
            Tags
          </label>
          <div
            class="c7"
            data-testid="tags-input"
          >
            <input
              class="c8"
              id="tags"
              name="tags"
              placeholder="Add tag and press space"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="c9"
      >
        <button
          class="c10"
          data-testid="primary-button"
          type="submit"
        >
          Add tool
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
